File: packages/disk/src/workspace.ts
import type {
  Disk,
  ExecResult,
  GrepMatch,
  GrepOptions,
  GrepResult,
  ListObjectsOptions,
  ListObjectsResult,
  PutObjectOptions,
  PutObjectResult,
  S3Object,
} from "./disk.js";
import type { ExecMount, ExecMountSpec, ExecOptions } from "./archil.js";
import type { FileSystem } from "./filesystem.js";
import { toSegments } from "./paths.js";

/** Anything that can run a multi-disk exec — i.e. an `Archil` client. A narrow
 * structural type so this module doesn't import the Archil class (avoiding a
 * value-import cycle with archil.ts). */
export interface WorkspaceExecCapable {
  exec(opts: ExecOptions): Promise<ExecResult>;
}

function isExecMountSpec(m: ExecMount): m is ExecMountSpec {
  return typeof m === "object" && m !== null && "disk" in m;
}

interface MountEntry {
  disk: Disk;
  readOnly: boolean;
  subdirectory?: string;
  conditional: boolean;
  queueMs?: number;
  checkoutPaths?: string[];
}

/**
 * A key-routed filesystem spanning several Archil disks. Each disk is mounted
 * under a name and addressed as the first segment of a key (`<name>/...`);
 * `getObject`, `putObject`, `deleteObject`, `listObjects`, and `grep` route to
 * the right disk by that segment (and `listObjects` / `grep` fan out across all
 * of them when the key/prefix doesn't name one). Implements the same
 * {@link FileSystem} surface as a single {@link Disk}, so it works anywhere a
 * disk does. Mounts can be changed at runtime with {@link addDisk} /
 * {@link removeDisk}.
 */
export class Workspace implements FileSystem {
  private readonly mounts = new Map<string, MountEntry>();

  /** @internal Use `archil.workspace({ ... })`. */
  constructor(
    private readonly client: WorkspaceExecCapable,
    mounts: Record<string, ExecMount>,
  ) {
    for (const [name, value] of Object.entries(mounts)) this.addDisk(name, value);
    if (this.mounts.size === 0) throw new Error("workspace() needs at least one disk");
  }

  // --- mount management --------------------------------------------------

  /** Mount (or replace) a disk at `name`; its objects are addressed as
   * `<name>/...`. Accepts a `Disk` or a mount spec (read-only / subdirectory /
   * conditional / delegation checkouts); a bare disk-id string is rejected —
   * fetch the disk first. */
  addDisk(name: string, disk: ExecMount): this {
    const rel = name.replace(/^\/+|\/+$/g, "");
    if (!rel) throw new Error("workspace mount name must be non-empty");
    // Routing is by the first key segment, so a mount name can't contain "/".
    if (rel.includes("/")) {
      throw new Error(`workspace mount name '${name}' must not contain '/'`);
    }
    // "." and ".." are path navigation to the router, not addressable labels.
    if (rel === "." || rel === "..") {
      throw new Error(`workspace mount name '${name}' is reserved ('.' and '..' are path navigation, not disk names)`);
    }
    let d: Disk | string;
    let readOnly = false;
    let subdirectory: string | undefined;
    let conditional = false;
    let queueMs: number | undefined;
    let checkoutPaths: string[] | undefined;
    if (isExecMountSpec(disk)) {
      d = disk.disk;
      readOnly = disk.readOnly ?? false;
      subdirectory = disk.subdirectory;
      conditional = disk.conditional ?? false;
      queueMs = disk.queueMs;
      checkoutPaths = disk.checkoutPaths;
    } else {
      d = disk;
    }
    if (typeof d === "string") {
      throw new Error(
        "workspace needs Disk objects, not disk-id strings; fetch with archil.getDisk(id) first",
      );
    }
    this.mounts.set(rel, {
      disk: d,
      readOnly,
      subdirectory,
      conditional,
      queueMs,
      checkoutPaths,
    });
    return this;
  }

  /** Unmount the disk at `name`. Returns whether a disk was removed. Refuses to
   * remove the last disk — a workspace must always have at least one (the same
   * invariant the constructor enforces), else fan-out/exec would have nothing to
   * route to. */
  removeDisk(name: string): boolean {
    const rel = name.replace(/^\/+|\/+$/g, "");
    if (this.mounts.has(rel) && this.mounts.size === 1) {
      throw new Error("cannot remove the last disk from a workspace; a workspace must keep at least one");
    }
    return this.mounts.delete(rel);
  }

  /** The names of the currently-mounted disks. */
  diskNames(): string[] {
    return [...this.mounts.keys()];
  }

  // --- routing -----------------------------------------------------------

  private unknownDisk(name: string): Error {
    return new Error(
      `No disk named '${name}'. Address objects by disk: ${this.diskNames().map((n) => `${n}/…`).join(", ")}`,
    );
  }

  /** Resolve a workspace key (`<name>/...`) to the disk and the disk-relative key. */
  private route(key: string): { entry: MountEntry; diskKey: string } {
    const segs = toSegments(key);
    if (segs.length === 0) {
      throw new Error("that key is the workspace root, not an object; name a disk, e.g. data/file.txt");
    }
    const entry = this.mounts.get(segs[0]);
    if (!entry) throw this.unknownDisk(segs[0]);
    return { entry, diskKey: this.diskKey(entry, segs.slice(1).join("/")) };
  }

  /** Mounts touched by a key prefix; an empty prefix fans out to all of them. */
  private covered(prefix: string): Array<{ name: string; entry: MountEntry; rel: string }> {
    const segs = toSegments(prefix);
    if (segs.length === 0) {
      return [...this.mounts.entries()].map(([name, entry]) => ({ name, entry, rel: this.diskKey(entry, "") }));
    }
    const entry = this.mounts.get(segs[0]);
    if (!entry) throw this.unknownDisk(segs[0]);
    return [{ name: segs[0], entry, rel: this.diskKey(entry, segs.slice(1).join("/")) }];
  }

  /** Map a workspace-relative key to the disk's key, applying the mount's subdirectory. */
  private diskKey(entry: MountEntry, rel: string): string {
    const trimmed = rel.replace(/^\/+/, "");
    const sub = (entry.subdirectory ?? "").replace(/^\/+|\/+$/g, "");
    if (!sub) return trimmed;
    return trimmed ? `${sub}/${trimmed}` : sub;
  }

  /** Map a disk key back to its workspace key (`<name>/...`), stripping the mount subdirectory. */
  private abs(name: string, entry: MountEntry, diskKey: string): string {
    const sub = (entry.subdirectory ?? "").replace(/^\/+|\/+$/g, "");
    let rel = diskKey;
    if (sub && diskKey.startsWith(sub + "/")) rel = diskKey.slice(sub.length + 1);
    else if (sub && diskKey === sub) rel = "";
    rel = rel.replace(/^\/+|\/+$/g, "");
    return rel ? `${name}/${rel}` : name;
  }

  // --- FileSystem --------------------------------------------------------

  async getObject(key: string): Promise<Uint8Array> {
    const { entry, diskKey } = this.route(key);
    return entry.disk.getObject(diskKey);
  }

  async putObject(
    key: string,
    body: string | Uint8Array | ArrayBuffer,
    options?: string | PutObjectOptions,
  ): Promise<PutObjectResult> {
    const { entry, diskKey } = this.route(key);
    if (entry.readOnly) throw new Error(`${key} is on a read-only disk and cannot be written.`);
    return entry.disk.putObject(diskKey, body, options);
  }

  async deleteObject(key: string): Promise<void> {
    const { entry, diskKey } = this.route(key);
    if (entry.readOnly) throw new Error(`${key} is on a read-only disk and cannot be deleted.`);
    await entry.disk.deleteObject(diskKey);
  }

  async listObjects(prefix?: string, opts: ListObjectsOptions = {}): Promise<ListObjectsResult> {
    // At the workspace root, the disks themselves are the top-level directories.
    // A non-recursive listing returns just those (like a delimited S3 listing),
    // rather than fanning out into each disk's contents — that only happens when
    // recursive is set or a prefix names a disk.
    if (toSegments(prefix ?? "").length === 0 && !opts.recursive) {
      return {
        objects: [],
        commonPrefixes: this.diskNames().map((n) => `${n}/`).sort(),
        isTruncated: false,
        keyCount: 0,
      };
    }
    const covered = this.covered(prefix ?? "");
    // Fan out resiliently: a disk that errors shouldn't sink the whole listing —
    // skip it and flag the result incomplete (isTruncated) rather than throwing.
    const listings = await Promise.allSettled(
      covered.map(({ entry, rel }) => {
        const p = rel ? (rel.endsWith("/") ? rel : `${rel}/`) : undefined;
        return entry.disk.listObjects(p, { recursive: opts.recursive });
      }),
    );
    const objects: S3Object[] = [];
    const commonPrefixes: string[] = [];
    let isTruncated = false;
    let keyCount = 0;
    covered.forEach(({ name, entry }, i) => {
      const settled = listings[i];
      if (settled.status === "rejected") {
        isTruncated = true; // a disk failed → the merged listing may be incomplete
        return;
      }
      const listing = settled.value;
      for (const cp of listing.commonPrefixes) {
        commonPrefixes.push(`${this.abs(name, entry, cp.replace(/\/+$/, ""))}/`);
      }
      for (const obj of listing.objects) {
        objects.push({ ...obj, key: this.abs(name, entry, obj.key) });
      }
      isTruncated = isTruncated || listing.isTruncated;
      keyCount += listing.keyCount;
    });
    objects.sort((a, b) => (a.key < b.key ? -1 : a.key > b.key ? 1 : 0));
    commonPrefixes.sort();
    return { objects, commonPrefixes, isTruncated, keyCount };
  }

  async grep(opts: GrepOptions): Promise<GrepResult> {
    const covered = this.covered(opts.directory ?? "");
    // Fan out resiliently: a disk that errors contributes a "list_failed" reason
    // (surfaced as a partial-results caveat) instead of failing the whole grep.
    const results = await Promise.allSettled(
      covered.map(({ entry, rel }) => entry.disk.grep({ ...opts, directory: rel })),
    );
    const matches: GrepMatch[] = [];
    let filesScanned = 0;
    let containersDispatched = 0;
    let computeSecondsUsed = 0;
    let durationMs = 0;
    let listingMs = 0;
    let grepMs = 0;
    const reasons = new Set<string>();
    covered.forEach(({ name, entry }, i) => {
      const settled = results[i];
      if (settled.status === "rejected") {
        reasons.add("list_failed");
        return;
      }
      const r = settled.value;
      filesScanned += r.filesScanned;
      containersDispatched += r.containersDispatched ?? 0;
      computeSecondsUsed += r.computeSecondsUsed ?? 0;
      durationMs = Math.max(durationMs, r.durationMs ?? 0);
      listingMs = Math.max(listingMs, r.listingMs ?? 0);
      grepMs = Math.max(grepMs, r.grepMs ?? 0);
      reasons.add(r.stoppedReason);
      for (const m of r.matches) matches.push({ file: this.abs(name, entry, m.file), line: m.line, text: m.text });
    });
    // Merging across disks can exceed the cap even when every disk completed, so
    // flag the truncation ourselves before slicing — otherwise the extra matches
    // are dropped with no signal.
    const cap = opts.maxResults ?? 1000;
    if (matches.length > cap) reasons.add("max_results");
    // Report the most serious reason so a listing/scan failure on one disk isn't
    // masked by a truncation reason on another.
    const priority = ["list_failed", "incomplete", "deadline", "max_results"];
    const stoppedReason = [...reasons].every((x) => x === "completed")
      ? "completed"
      : (priority.find((x) => reasons.has(x)) ?? "completed");
    return {
      matches: matches.slice(0, cap),
      stoppedReason,
      filesScanned,
      containersDispatched,
      computeSecondsUsed,
      durationMs,
      listingMs,
      grepMs,
    } as GrepResult;
  }

  exec(command: string): Promise<ExecResult> {
    const disks: Record<string, ExecMount> = {};
    for (const [name, entry] of this.mounts) {
      const hasMountOptions =
        entry.readOnly ||
        entry.subdirectory ||
        entry.conditional ||
        entry.queueMs !== undefined ||
        entry.checkoutPaths !== undefined;
      disks[name] = hasMountOptions
        ? {
            disk: entry.disk,
            subdirectory: entry.subdirectory,
            readOnly: entry.readOnly,
            conditional: entry.conditional,
            queueMs: entry.queueMs,
            checkoutPaths: entry.checkoutPaths,
          }
        : entry.disk;
    }
    return this.client.exec({ disks, command });
  }
}

File: python/archil/_workspace.py
"""A key-routed filesystem spanning several Archil disks.

Each disk is mounted under a name and addressed as the first segment of a key
(``<name>/...``); ``get_object``, ``put_object``, ``delete_object``,
``list_objects``, and ``grep`` route to the right disk by that segment (and
``list_objects`` / ``grep`` fan out across all of them when the key/prefix
doesn't name one). Implements the same :class:`FileSystem` surface as a single
:class:`Disk` — using the very same method names — so it works anywhere a disk
does, not just with the agent tools. Mounts can be changed at runtime with
:meth:`add_disk` / :meth:`remove_disk`.

Like the rest of the SDK this is a synchronicity-wrapped impl: it holds the
*impl* disks and client (normalized via ``translate_in``) so its async methods
await them directly on the shared loop.
"""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import List, Optional

from archil_openapi.models.exec_disk_result import ExecDiskResult as ExecResult
from archil_openapi.models.grep_disk_result import GrepDiskResult as GrepResult
from archil_openapi.models.grep_match import GrepMatch
from archil_openapi.models.grep_stopped_reason import GrepStoppedReason

from ._http import BodyType
from ._models import ListObjectsResult, PutObjectResult, S3Object
from ._paths import to_segments
from ._synchronizer import translate_in, translate_out

# Imported at runtime (not under TYPE_CHECKING) so the synchronicity stub
# generator can resolve agent_tools()'s return type. agent_tools does not import
# _workspace, so this does not create an import cycle.
from .agent_tools import AgentToolset


@dataclass
class _Mount:
    disk: object  # impl _Disk
    read_only: bool = False
    subdirectory: Optional[str] = None
    conditional: bool = False


class _Workspace:
    """Use ``archil.workspace({...})`` to build one."""

    def __init__(self, client: object, mounts: dict) -> None:
        self._client = client  # impl _Archil
        self._mounts: dict[str, _Mount] = {}
        for name, value in mounts.items():
            self.add_disk(name, value)
        if not self._mounts:
            raise ValueError("workspace() needs at least one disk")

    # --- mount management --------------------------------------------------

    def add_disk(self, name: str, disk: object) -> "_Workspace":
        """Mount (or replace) a disk at ``name``; its objects are addressed as
        ``<name>/...``. Accepts a ``Disk`` or an ``ExecMountSpec`` (read-only /
        subdirectory / conditional); a bare disk-id string is rejected — fetch
        the disk first."""
        from ._archil import ExecMountSpec

        rel = name.strip("/")
        if not rel:
            raise ValueError("workspace mount name must be non-empty")
        # Routing is by the first key segment, so a mount name can't contain "/".
        if "/" in rel:
            raise ValueError(f"workspace mount name {name!r} must not contain '/'")
        # "." and ".." are path navigation to the router, not addressable labels.
        if rel in (".", ".."):
            raise ValueError(
                f"workspace mount name {name!r} is reserved ('.' and '..' are path navigation, not disk names)"
            )
        if isinstance(disk, ExecMountSpec):
            d, read_only, subdirectory, conditional = (
                disk.disk, disk.read_only, disk.subdirectory, disk.conditional
            )
        else:
            d, read_only, subdirectory, conditional = disk, False, None, False
        if isinstance(d, str):
            raise ValueError(
                "workspace needs Disk objects, not disk-id strings; "
                "fetch the disk with archil.get_disk(id) first"
            )
        self._mounts[rel] = _Mount(
            disk=translate_in(d), read_only=read_only, subdirectory=subdirectory, conditional=conditional
        )
        return self

    def remove_disk(self, name: str) -> bool:
        """Unmount the disk at ``name``. Returns whether a disk was removed.
        Refuses to remove the last disk — a workspace must always have at least
        one (the same invariant the constructor enforces), else fan-out/exec
        would have nothing to route to."""
        rel = name.strip("/")
        if rel in self._mounts and len(self._mounts) == 1:
            raise ValueError(
                "cannot remove the last disk from a workspace; a workspace must keep at least one"
            )
        return self._mounts.pop(rel, None) is not None

    def disk_names(self) -> List[str]:
        """The names of the currently-mounted disks."""
        return list(self._mounts)

    # --- routing -----------------------------------------------------------

    def _unknown_disk(self, name: str) -> ValueError:
        return ValueError(
            f"No disk named {name!r}. Address objects by disk: "
            + ", ".join(f"{n}/…" for n in self._mounts)
        )

    def _route(self, key: str) -> tuple[_Mount, str]:
        segs = to_segments(key)
        if not segs:
            raise ValueError("that key is the workspace root, not an object; name a disk, e.g. data/file.txt")
        entry = self._mounts.get(segs[0])
        if entry is None:
            raise self._unknown_disk(segs[0])
        return entry, self._disk_key(entry, "/".join(segs[1:]))

    def _covered(self, prefix: str) -> list[tuple[str, _Mount, str]]:
        """Mounts touched by a key prefix; an empty prefix fans out to all of them."""
        segs = to_segments(prefix)
        if not segs:
            return [(name, entry, self._disk_key(entry, "")) for name, entry in self._mounts.items()]
        entry = self._mounts.get(segs[0])
        if entry is None:
            raise self._unknown_disk(segs[0])
        return [(segs[0], entry, self._disk_key(entry, "/".join(segs[1:])))]

    def _disk_key(self, entry: _Mount, rel: str) -> str:
        rel = rel.lstrip("/")
        sub = (entry.subdirectory or "").strip("/")
        if not sub:
            return rel
        return f"{sub}/{rel}" if rel else sub

    def _abs(self, name: str, entry: _Mount, disk_key: str) -> str:
        sub = (entry.subdirectory or "").strip("/")
        rel = disk_key
        if sub and disk_key.startswith(sub + "/"):
            rel = disk_key[len(sub) + 1:]
        elif sub and disk_key == sub:
            rel = ""
        rel = rel.strip("/")
        return f"{name}/{rel}" if rel else name

    # --- FileSystem --------------------------------------------------------

    async def get_object(self, key: str) -> bytes:
        entry, disk_key = self._route(key)
        return await entry.disk.get_object(disk_key)

    async def put_object(
        self,
        key: str,
        body: BodyType,
        content_type: Optional[str] = None,
        *,
        mode: Optional[int] = None,
        uid: Optional[int] = None,
        gid: Optional[int] = None,
    ) -> PutObjectResult:
        entry, disk_key = self._route(key)
        if entry.read_only:
            raise ValueError(f"{key} is on a read-only disk and cannot be written.")
        return await entry.disk.put_object(
            disk_key, body, content_type, mode=mode, uid=uid, gid=gid
        )

    async def delete_object(self, key: str) -> None:
        entry, disk_key = self._route(key)
        if entry.read_only:
            raise ValueError(f"{key} is on a read-only disk and cannot be deleted.")
        await entry.disk.delete_object(disk_key)

    async def list_objects(self, prefix: Optional[str] = None, *, recursive: bool = False) -> ListObjectsResult:
        # At the workspace root, the disks themselves are the top-level directories.
        # A non-recursive listing returns just those (like a delimited S3 listing),
        # rather than fanning out into each disk's contents — that only happens when
        # recursive is set or a prefix names a disk.
        if not to_segments(prefix or "") and not recursive:
            return ListObjectsResult(
                objects=[],
                common_prefixes=sorted(f"{n}/" for n in self._mounts),
                is_truncated=False,
                key_count=0,
                prefix=prefix,
            )
        covered = self._covered(prefix or "")
        # Fan out resiliently: a disk that errors shouldn't sink the whole listing
        # — skip it and flag the result incomplete (is_truncated) rather than raise.
        listings = await asyncio.gather(
            *(self._list_one(entry, rel, recursive) for _, entry, rel in covered),
            return_exceptions=True,
        )
        objects: List[S3Object] = []
        common_prefixes: List[str] = []
        is_truncated = False
        key_count = 0
        for (name, entry, _), listing in zip(covered, listings):
            if isinstance(listing, BaseException):
                is_truncated = True  # a disk failed → the merged listing may be incomplete
                continue
            for cp in listing.common_prefixes:
                common_prefixes.append(f"{self._abs(name, entry, cp.rstrip('/'))}/")
            for obj in listing.objects:
                objects.append(
                    S3Object(
                        key=self._abs(name, entry, obj.key),
                        size=obj.size,
                        etag=obj.etag,
                        last_modified=obj.last_modified,
                    )
                )
            is_truncated = is_truncated or listing.is_truncated
            key_count += listing.key_count
        objects.sort(key=lambda o: o.key)
        common_prefixes.sort()
        return ListObjectsResult(
            objects=objects,
            common_prefixes=common_prefixes,
            is_truncated=is_truncated,
            key_count=key_count,
            prefix=prefix,
        )

    async def _list_one(self, entry: _Mount, rel: str, recursive: bool):
        disk_prefix = f"{rel}/" if rel and not rel.endswith("/") else (rel or None)
        return await entry.disk.list_objects(disk_prefix, recursive=recursive)

    async def grep(
        self,
        *,
        directory: str,
        pattern: str,
        recursive: bool = False,
        max_duration_seconds: int = 30,
        concurrency: int = 50,
        max_results: int = 1000,
    ) -> GrepResult:
        covered = self._covered(directory)
        # Fan out resiliently: a disk that errors contributes a "list_failed"
        # reason (surfaced as a partial-results caveat) instead of failing the grep.
        results = await asyncio.gather(
            *(
                entry.disk.grep(
                    directory=rel,
                    pattern=pattern,
                    recursive=recursive,
                    max_duration_seconds=max_duration_seconds,
                    concurrency=concurrency,
                    max_results=max_results,
                )
                for _, entry, rel in covered
            ),
            return_exceptions=True,
        )
        matches: List[GrepMatch] = []
        files_scanned = containers_dispatched = duration_ms = listing_ms = grep_ms = 0
        compute_seconds_used = 0.0
        reasons: set[str] = set()
        for (name, entry, _), r in zip(covered, results):
            if isinstance(r, BaseException):
                reasons.add("list_failed")
                continue
            files_scanned += r.files_scanned
            containers_dispatched += r.containers_dispatched
            compute_seconds_used += r.compute_seconds_used
            duration_ms = max(duration_ms, r.duration_ms)
            listing_ms = max(listing_ms, r.listing_ms)
            grep_ms = max(grep_ms, r.grep_ms)
            reasons.add(r.stopped_reason)
            for m in r.matches:
                matches.append(GrepMatch(self._abs(name, entry, m.file), m.line, m.text))
        # Merging across disks can exceed the cap even when every disk completed,
        # so flag the truncation ourselves before slicing — otherwise the extra
        # matches are dropped with no signal.
        if len(matches) > max_results:
            reasons.add("max_results")
        # Report the most serious reason so a listing/scan failure on one disk
        # isn't masked by a truncation reason on another.
        priority = ("list_failed", "incomplete", "deadline", "max_results")
        stopped_reason = (
            "completed" if reasons <= {"completed"}
            else next((r for r in priority if r in reasons), "completed")
        )
        return GrepResult(
            matches=matches[:max_results],
            stopped_reason=GrepStoppedReason(stopped_reason),
            files_scanned=files_scanned,
            containers_dispatched=containers_dispatched,
            compute_seconds_used=compute_seconds_used,
            duration_ms=duration_ms,
            listing_ms=listing_ms,
            grep_ms=grep_ms,
        )

    async def exec(self, command: str) -> ExecResult:
        from ._archil import ExecMountSpec

        disks: dict = {}
        for name, entry in self._mounts.items():
            if entry.read_only or entry.subdirectory or entry.conditional:
                disks[name] = ExecMountSpec(
                    disk=entry.disk,
                    subdirectory=entry.subdirectory,
                    read_only=entry.read_only,
                    conditional=entry.conditional,
                )
            else:
                disks[name] = entry.disk
        return await self._client.exec(disks=disks, command=command)

    # --- agent tools -------------------------------------------------------

    def agent_tools(self, *, tools: Optional[List[str]] = None) -> "AgentToolset":
        """Build a filesystem toolset for this workspace that drops into popular
        agent frameworks, exactly like :meth:`Disk.agent_tools` — but operations
        route to the right disk by key and ``grep`` / ``list_objects`` fan out
        across all of them.

            agent = Agent(tools=workspace.agent_tools().for_openai_agents())
        """
        from .agent_tools._toolset import AgentToolset

        return AgentToolset(translate_out(self), tools)
